### PR TITLE
Bedsheet Improvements

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request

This is a remake of #68602 that does:

- Adds context screentips for bedsheets
- Lets you attack other mobs that are lying down to cover them in a bedsheet
- Removes deprecated bedsheet code
- Fixes tucked in items (plushies, nuke disk, bedsheets) to use proper direction and rotation for beds
- Fixes bedsheets covering mobs in the wrong direction
- Changes bedsheets to only cover mobs that are lying down
- Bedsheets can now be rotated with AltClick

## Why It's Good For The Game

Easier to use and looks nice.

## Changelog

:cl:
qol: Bedsheets now have context screentips and will cover up mobs if they are attacked and lying down. Bedsheet can also be rotated with AltClick.
fix: Fix rotation and offsets to work properly on mobs, nuke disk, plushies, and bedsheets when placing something on a bed.
/:cl:

### CODE REVIEW

File: code/datums/elements/bed_tucking.dm [hunk 1]
```diff
@@ -56,5 +58,5 @@
 /datum/element/bed_tuckable/proc/untuck(obj/item/tucked)
 	SIGNAL_HANDLER
 
-	tucked.transform = turn(tucked.transform, -rotation_degree)
+	tucked.transform = turn(matrix(), 0) // reset rotation
```
Comment (Mothblocks) [inline code/datums/elements/bed_tucking.dm:61]: Is `turn(matrix(), 0)` different from just `matrix()`?

And does this not also destroy scale? What happens if I have a scale transformed object that I tuck in bed?

File: code/game/objects/structures/bedsheet_bin.dm [hunk 2]
```diff
@@ -37,31 +37,49 @@ LINEN BINS
...
+	if(!user.CanReach(target))
+		return
+	if(user.combat_mode)
+		return ..()
+	if(target.body_position != LYING_DOWN)
+		return
+	if(!user.dropItemToGround(src))
+		return
+
+	src.forceMove(get_turf(target))
```
Comment (Mothblocks) [inline code/game/objects/structures/bedsheet_bin.dm:59]:
```suggestion
	forceMove(get_turf(target))
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/elements/bed_tucking.dm (hunk 1 above)
Comment (timothymtorres) [reply code/datums/elements/bed_tucking.dm]: Fixed with https://github.com/tgstation/tgstation/pull/70576/commits/b6632982b312d567222873f5d0d7c4035c51eade

File: code/game/objects/structures/bedsheet_bin.dm [hunk 3]
```diff
@@ -37,31 +37,50 @@ LINEN BINS
...
-		else    //The player is rotated to the left, lay the sheet right!
-			dir = EAST
+
+	coverup(user)
 	add_fingerprint(user)
-	return
+
+/obj/item/bedsheet/proc/coverup(mob/living/sleeper)
+	layer = ABOVE_MOB_LAYER
+	SET_PLANE_IMPLICIT(src, GAME_PLANE_UPPER)
```
Comment (ShizCalev) [inline code/game/objects/structures/bedsheet_bin.dm:78]: register COMSIG_MOVABLE_MOVED on the sleeper to set the layer back to initial once they're gone?

Comment (timothymtorres) [reply code/game/objects/structures/bedsheet_bin.dm:78]: Done with https://github.com/tgstation/tgstation/pull/70576/commits/415291adb27136c4fa936080c022858419e12349

File: code/game/objects/structures/bedsheet_bin.dm [hunk 4]
```diff
@@ -75,6 +112,12 @@ LINEN BINS
 	else
 		return ..()
 
+/obj/item/bedsheet/AltClick(mob/living/user)
+	// double check the canUseTopic args to make sure it's correct
+	if(!istype(user) || !user.canUseTopic(src, be_close = TRUE, no_dexterity = TRUE, no_tk = FALSE, need_hands = !iscyborg(user)))
+		return
+	dir = dir & WEST ? EAST : WEST
```
Comment (JohnFulpWillard) [inline code/game/objects/structures/bedsheet_bin.dm:119]: I'm not really a fan of this, do you think it would be better to just turn() it 180?

Comment (timothymtorres) [reply code/game/objects/structures/bedsheet_bin.dm]: Applied with https://github.com/tgstation/tgstation/pull/70576/commits/9535a90ffe959db5fab4443642954a151ba0758b

File: code/game/objects/structures/bedsheet_bin.dm [hunk 5]
```diff
@@ -37,31 +37,68 @@ LINEN BINS
...
+
+	return NONE
+
+/obj/item/bedsheet/attack(mob/living/target, mob/living/user)
+	if(!user.CanReach(target))
+		return
+	if(user.combat_mode)
+		return ..()
+	if(target.body_position != LYING_DOWN)
+		return
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/game/objects/structures/bedsheet_bin.dm]: Applied with f45c1ecd38140d4c5265f22c68578b65870fbcf6

File: code/game/objects/structures/bedsheet_bin.dm [hunk 6]
```diff
@@ -37,31 +37,66 @@ LINEN BINS
...
-	return
+
+/obj/item/bedsheet/proc/coverup(mob/living/sleeper)
+	layer = ABOVE_MOB_LAYER
+	SET_PLANE_IMPLICIT(src, GAME_PLANE_UPPER)
+	pixel_x = 0
+	pixel_y = 0
+	balloon_alert(sleeper, "covered")
+	var/angle = sleeper.lying_prev
+	dir = angle == 90 ? WEST : EAST
```
Comment (JohnFulpWillard) [inline code/game/objects/structures/bedsheet_bin.dm:85]: turn() here too, sorry

Comment (timothymtorres) [reply code/game/objects/structures/bedsheet_bin.dm]: Applied with https://github.com/tgstation/tgstation/pull/70576/commits/936dc6ed13fc7d10343701c9f9b83f566ab0b0b4

File: code/game/objects/structures/bedsheet_bin.dm [hunk 7]
```diff
@@ -37,31 +37,68 @@ LINEN BINS
...
 		dying_key = DYE_REGISTRY_DOUBLE_BEDSHEET
+	register_item_context()
+
+/obj/item/bedsheet/examine(mob/user)
+	. = ..()
+	. += span_notice("The bedsheet looks like it can be <b>cut</b>.")
+	. += span_notice("The bedsheet can be rotated using AltClick.")
+
+/obj/item/bedsheet/add_item_context(datum/source, list/context, mob/living/target)
+	if(isliving(target) && target.body_position == LYING_DOWN)
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/game/objects/structures/bedsheet_bin.dm:52]: It should be possible? They are done in the same process.

https://github.com/tgstation/tgstation/blob/4d6a8bc5371eb61f9adb7c4dd1b8c441d3ae9251/code/game/atoms.dm#L1984-L1985

[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/game/objects/structures/bedsheet_bin.dm:55]: On second try, it works!  Applied with https://github.com/tgstation/tgstation/pull/70576/commits/abf00e6669d9f1f255d4fa4f60e93963526bd205

I think I forgot to include both of these when I initially attempted it:

```
register_context()
register_item_context()
```

I also couldn't find any examples of items having both types when I looked.  Regardless it's nice to know this is possible so big thanks!

**Pull request merged**